x-pack/metricbeat/module/aws/s3/s3.go

```diff
@@ -0,0 +1,302 @@
+		s3Logger.Error(err)
+	}
+
+	metricSet, err := aws.NewMetricSet(base)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating aws metricset")
+	}
+
+	// Get a list of regions
+	awsConfig := defaults.Config()
+	awsCreds := awssdk.Credentials{
+		AccessKeyID:     moduleConfig.AccessKeyID,
+		SecretAccessKey: moduleConfig.SecretAccessKey,
+	}
+	if moduleConfig.SessionToken != "" {
+		awsCreds.SessionToken = moduleConfig.SessionToken
+	}
+
+	awsConfig.Credentials = awssdk.StaticCredentialsProvider{
+		Value: awsCreds,
+	}
+
+	awsConfig.Region = moduleConfig.DefaultRegion
+
+	awsConfig.Region = moduleConfig.DefaultRegion
+	svcEC2 := ec2sdk.New(awsConfig)
+	regionsList, err := ec2.GetRegions(svcEC2)
+	if err != nil {
+		err = errors.Wrap(err, "GetRegions failed")
+		s3Logger.Error(err.Error())
+	}
+
+	// Calculate duration based on period
+	durationString, periodSec, err := ec2.ConvertPeriodToDuration(moduleConfig.Period)
+	if err != nil {
+		s3Logger.Error(err.Error())
+		return nil, err
+	}
+
+	// Check if period is set to be multiple of 60s or 300s
+	remainder300 := periodSec % 300
+	remainder60 := periodSec % 60
+	if remainder300 != 0 || remainder60 != 0 {
+		err := errors.New("period needs to be set to 60s (or a multiple of 60s) or set to 300s " +
+			"(or a multiple of 300s). To avoid data missing or extra costs, please make sure period is set correctly " +
+			"in config.yml")
+		s3Logger.Info(err)
+	}
+
+	return &MetricSet{
+		MetricSet:      metricSet,
+		moduleConfig:   &moduleConfig,
+		awsConfig:      &awsConfig,
+		regionsList:    regionsList,
+		durationString: durationString,
+		periodInSec:    periodSec,
+		logger:         s3Logger,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) {
+	namespace := "AWS/S3"
+	// Get startTime and endTime
+	startTime, endTime, err := getStartTimeEndTime(m.durationString)
+	if err != nil {
+		logp.Error(errors.Wrap(err, "Error ParseDuration"))
+		m.logger.Error(err.Error())
+		report.Error(err)
+		return
+	}
+
+	// GetMetricData for AWS S3 from Cloudwatch
+	for _, regionName := range m.regionsList {
+		m.awsConfig.Region = regionName
+		svcCloudwatch := cloudwatch.New(*m.awsConfig)
+		listMetricsInput := &cloudwatch.ListMetricsInput{Namespace: &namespace}
+		reqListMetrics := svcCloudwatch.ListMetricsRequest(listMetricsInput)
+
+		// List metrics of S3 for each region
+		listMetricsOutput, err := reqListMetrics.Send()
+		if err != nil {
+			err = errors.Wrap(err, "ListMetricsRequest failed, skipping region "+regionName)
+			m.logger.Error(err.Error())
+			report.Error(err)
+			continue
+		}
+
+		if listMetricsOutput.Metrics == nil || len(listMetricsOutput.Metrics) == 0 {
+			m.logger.Info("No S3 buckets in region " + regionName)
```

Comment:
If I have S3 buckets only in 1 region, I get this logged for all other regions on every fetch?